src/SAML2/AuthnRequest.php

```diff
@@ -307,15 +305,40 @@ public function buildRequestQuery()
      */
     public function getSignedRequestQuery()
```

Comment:
:christmas_tree: **Ignore this message if it's not yet 2017 and you're enjoying your days off!** :christmas_tree: 

---

I think it would be good to unit-test this method and/or extract the query parameter parsing algorithm and unit-test that.

Also, I woke up this morning wondering about how we parse the query parameters in this application (is this healthy? :thinking:) in both RedirectBinding and this method, and whether there should be one canonical way of reading and verifying a message that is received through the RedirectBinding. The RedirectBinding uses Symfony\Request to extract query parameters, and this method performs manual query parameter extraction. If ever the two methods deviate regarding parameter order, this could _potentially_ introduce an attack vector where we (for example) verify the signature of one `SAMLRequest` parameter (say, the first) and then handle the request that is present in the last `SAMLRequest` query parameter. Currently both methods use the last query parameter, and ignore any previous ones, so this is _not_ the case, currently. The point is that these things can go askew without us noticing it.

Concretely:

1. We may need one canonical query parsing and reading algorithm for all SAML communication to preclude any future deviations in parameter order, encoding, etc. etc. that may unexpectedly appear between Symfony\Request and the method used below.
1. We should test that this system does not allow verification of one SAMLRequest and handling of another due to confusion of same-named query parameter.
1. We should test that our RedirectBinding does not allow duplicate parameters that are relevant to the [RedirectBinding message](https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf).

I think point 1 is important for this PR. The other points could be addressed in separate issues.